### ISSUE TITLE
s3 client cannot head or get objection versions for delete markers

### ISSUE THREAD

**Issue by jstewmon** [boto/botocore#674]
The s3 api returns a 405 status code for HEAD and GET requests using a versionId for a delete marker, which causes botocore to raise a `ClientError` (in `BaseClient._make_api_call`).

The docs indicate that both `head_object` and `get_object` should return a `dict` with `DeleteMarker: True` when the requested `VersionId` is for a delete marker.

This seems like it could actually be a bug in S3, which should probably not respond with a 405 for either method when a `versionId` is supplied.

**Comment by rayluo** [comment]
Thank you @jstewmon . This is a great question because of its depth. After (quite) some investigation, I believe what you observed is actually the correct behavior. And here comes the detailed explanation.

1. The protocol on the wire used by S3 is RESTful. The HTTP 405 error alone, clearly indicates that, S3 does NOT accept HEAD or GET method for the delete marker.  We can even double confirm (although not really necessary) this from S3 documentation, ["The only operation you can use on a delete marker is DELETE"](http://docs.aws.amazon.com/AmazonS3/latest/dev/DeleteMarker.html).

2. Now the documentation part. I have to admit that the [doc in Botocore](http://botocore.readthedocs.org/en/latest/reference/services/s3.html#S3.Client.get_object) and the [doc in Boto 3](http://boto3.readthedocs.org/en/latest/reference/services/s3.html#S3.Object.get) are misleading to make you (and me too) think that there would be a DeleteMarker key in the GET response dictionary. But the fact is, there won't be such a response at all, because an exception will be raised.

3. So the next question would be, shall we adjust the documentation in this case?  Unfortunately, the answer seems to be "sorry there is no easy way to do it". The entire Boto 3 and Botocore infrastructure are designed to be [data-driven](http://botocore.readthedocs.org/en/latest/tutorial/index.html).  The delete marker description of GET operation in the two documentation pages I pasted above, plus the relevant [documentation in AWS CLI](http://docs.aws.amazon.com/cli/latest/reference/s3api/get-object.html), are all derived from one source, the [S3's API documentation](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html). We don't have bandwidth to maintain them manually.

In short, do not attempt to GET or HEAD a delete marker. You can simply delete it when you want to. And do not expect there will be a DeleteMarker "attribute" actually exists in the Botocore or Boto3 API response.

Hope that helps.

**Comment by jstewmon** [comment]
Thanks for the detailed explanation. Those are unfortunate circumstances for the documentation.

I can appreciate that most of the documentation must be generated from upstream data sources, but it would be great if sdk-specific remarks could be added to help clarify situations like this.

I stumbled into this issue trying to use boto3's `Bucket.object_versions` resource collection to list all versions of an object. That interface returns `ObjectVersion` instances for both `Versions` and `DeleteMarkers`, but there is no property on `ObjectVersion` to indicate whether the instance represents a `DeleteMarker` or a `Version`. It turns out that the only way to identify an `ObjectVersion` as a `DeleteMarker` is to check if `Size` or `ETag`  is `None` (assuming there is no other case when those attributes can be `None`).

It looks like boto3 relies on the shapes defined in botocore for its resource types, so I suppose it is not possible to define a better resource type for ObjectVersion that has an `is_delete_marker` property, which the resource collection could populate?

As far as I can tell, the only clean way of identifying delete markers in boto3 is to use the `list_object_versions` paginator, handling the `DeleteMarkers` and `Versions` attributes oneself. Is that right?

**Comment by rayluo** [comment]
Thanks for your investigation. I am adding some minor comments based on your report.

Boto3's `Bucket.object_versions` resource collection to list all versions of an object. That interface returns `ObjectVersion` instances for both `Versions` and `DeleteMarkers`, but there is no property on `ObjectVersion` to indicate whether the instance represents a `DeleteMarker` or a `Version`. It turns out that the only way to identify an `ObjectVersion` as a `DeleteMarker` is to check if `Size` or `ETag` **or `StorageClass`**  is ~~`None` (assuming there is no other case when those attributes can be `None`)~~ **absent**.

> It looks like boto3 relies on the shapes defined in botocore for its resource types, so I suppose it is not possible to define a better resource type for ObjectVersion that has an `is_delete_marker` property, which the resource collection could populate?

It seems so.

> As far as I can tell, the only clean way of identifying delete markers in boto3 is to use the `list_object_versions` paginator, handling the `DeleteMarkers` and `Versions` attributes oneself. Is that right?

That is also a viable option.
